Entities

cdworak opened the pull request:
this isn't even really a pull request because it's not done, just so hup can see

mhupman commented on client.go:
```diff
@@ -324,7 +325,7 @@ func tokenListHelper(lr tokenListRetriever, opts *ListOptions) error {
 	u, err := url.Parse(requrl)
 	if err != nil {
-		return "", nil
+		return "", err
```
and a bugfix - hurrah!

mhupman commented on client.go:
```diff
@@ -336,6 +337,7 @@ func addListOptions(requrl string, opts *ListOptions) (string, error) {
 		q.Add("limit", strconv.Itoa(opts.Limit))
 	}
 	if opts.PageToken != "" {
+		// TODO: this is not compatible with the old parameter name
```
:(

maybe the func could take another parameter if the new/old apis are the same in everything other than param name

mhupman commented on customfield.go:
```diff
@@ -231,18 +233,18 @@ func (v *VideoGallery) CustomFieldTag() string {
 
 type DailyTimes struct {
-	DailyTimes string `json:"dailyTimes,omitempty"`
+	Sunday    string `json:"sunday,omitempty"`
```
was this wrong all along or a change in Entities?

cdworak replied:
a change in Entities...I copied over the old struct. same thing happened for hours. will figure out how to distinguish between the two

mhupman commented on customfield_service.go:
```diff
@@ -531,6 +531,7 @@ func ParseCustomFields(cfraw map[string]interface{}, cfs []*CustomField) (map[st
 	return parsed, nil
 }
 
+// TODO: Can we do this anymore?
```
probably not, and it probably doesn't make sense.  it was just to prevent the "obvious" mistake of passing in a string vs id.   now we probably won't know until we send to the API (unless we validate against the field schema, but that is probably a new exposed method, not an automagic thing)

mhupman commented on customfield_service_test.go:
```diff
@@ -42,11 +42,11 @@ func runParseTest(t *testing.T, index int, c customFieldParseTest) {
 
 	if cfType != expectType {
-		t.Errorf("test #%d (%s) failed: expected type %s, got type %s", index, c.TypeName, expectType, cfType)
+		t.Errorf("test #%d (%s) failed:\nexpected type %s\ngot type %s", index, c.TypeName, expectType, cfType)
```
i really hate how hard it is to format things in test output

mhupman replied:
the tabbing in particular

mhupman commented on customfield_service_test.go:
```diff
@@ -59,15 +59,58 @@ var (
-				"date":  "2016-05-31",
-				"hours": "9:00:17:00",
+				"date": "2016-05-31",
+				"hours": []*Times{
```
you chose interface types above but a concrete type here - if you're trying to mimic json should be primitive values, like maps and strings?

mhupman replied:
(the reason being that Times could have its own MarshalJSON define which would mess up your tests)

cdworak replied:
yep will fix

mhupman commented on entity.go:
```diff
@@ -0,0 +1,22 @@
+type EntityType string
+
+// This is a little dangerous because if you embed an Entity (like Location) within another Entity
+// You don't have to re-define these...but we'd need to re-define copy...
```
comment mentions "copy"

mhupman commented on entity.go:
```diff
@@ -0,0 +1,22 @@
+
+// This is a little dangerous because if you embed an Entity (like Location) within another Entity
+// You don't have to re-define these...but we'd need to re-define copy...
+type Entity interface {
```
love the narrow interface!

mhupman commented on entity.go:
```diff
@@ -0,0 +1,22 @@
+	CategoryIds *[]string         `json:"categoryIds,omitempty"`
+	Language    *string           `json:"language,omitempty"`
+	CountryCode *string           `json:"countryCode,omitempty"`
+	nilIsEmpty  bool
```
oh no do we still need this thing :(

cdworak replied:
I'm not sure yet but don't want to forget about it

mhupman commented on entity.go:
```diff
@@ -0,0 +1,22 @@
+package yext
```
i might suggest adding an Entity type here that is appropriate as an Unmarshal target for any entity... basically just 

type BaseEntity struct {
 Meta *EntityMeta `json:"meta"`
}

maybe no other attributes.  it would also satisfy the Entity interface

cdworak replied:
With the idea being that we would embed this guy in other Entities like Location? Or just have him around as the most stripped down entity?

mhupman replied:
yep, could embed to get Meta "for free" without need to duplicate.  it would also be a good deser target if you didn't know what you were going to get from the API - for example, if you hit the generic entities endpoint to return any and all entities (assuming that exists?)?

cdworak replied:
got it

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+func (e *EntityService) LookupEntity(entityType EntityType) (interface{}, error) {
+	val, ok := e.registry[entityType]
+	if !ok {
+		return nil, fmt.Errorf("Unable to find entity type %s in entity registry %v", entityType, e.registry)
```
instead of spitting out the whole map, maybe just print out the names of the "known" types?

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+	return reflect.New(reflect.TypeOf(val)).Interface(), nil
+}
+
+func GetBytes(key interface{}) ([]byte, error) {
```
heh too much copy/pasta

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+	e.registry[entityType] = newVal
+}
+
+func (e *EntityService) LookupEntity(entityType EntityType) (interface{}, error) {
```
this doesn't just lookup but it also creates an instance of the type, right?  might adjust the name

mhupman commented on entity_test.go:
```diff
@@ -0,0 +1,356 @@
+}
+
+func (l *CustomLocationEntity) EntityId() string {
+	return ""
```
?

mhupman commented on entity_test.go:
```diff
@@ -0,0 +1,356 @@
+}
+
+func (l *CustomLocationEntity) Type() EntityType {
+	return ENTITYTYPE_LOCATION
```
?

mhupman replied:
can we get this info off of Meta, which is found in the embedded Location?

cdworak replied:
I think i was worried that if someone creates a new Location type and doesn't set the type...then returning EntityType would give you an ""

But I suppose the same thing would happen if you don't set the Id so you would just need to not forget?

mhupman replied:
Oh, you mean on create?  If its pulled from the API, it would always be populated for you, right?

mhupman commented on entity_test.go:
```diff
@@ -0,0 +1,356 @@
+	return locationsPath
+}
+
+func (l *CustomLocationEntity) Copy() Entity {
```
do we need this?

mhupman commented on error.go:
```diff
@@ -101,3 +102,54 @@ func IsErrorCode(err error, code int) bool {
+	return &Error{Type: typ, Code: codeInt, Message: message}, nil
+}
+
+func ErrorsFromString(errorStr string) ([]*Error, error) {
```
why do we need to go backwards from the printed version of an error?

cdworak replied:
I didn't write these....let me look into why they popped up all of a sudden

mhupman replied:
whew!

mhupman commented on event.go:
```diff
@@ -0,0 +1,34 @@
+package yext
```
suggestion - prefix the filenames of these things so we know they go together - `entity_event.go`, for example

cdworak replied:
I think this is some of the interns work for the summer that somehow got on this PR. i'll remove it

mhupman replied:
😱

mhupman replied:
bump on removal

mhupman commented on hydrate.go:
```diff
@@ -2,7 +2,7 @@ package yext
 import "fmt"
 
-// TODO: This mutates the location, no need to return the value
+// TODO: I think this can be deleted (unless we keep the location-service around)
```
FWIW i think we have to keep location service around as long as it remains in the API

mhupman commented on language_profile_service.go:
```diff
@@ -31,9 +31,9 @@ func (l *LanguageProfileService) GetAll(id string) (*LanguageProfileListResponse
-	if _, err := l.HydrateLocations(v.LanguageProfiles); err != nil {
-		return nil, r, err
-	}
+	// if _, err := l.HydrateLocations(v.LanguageProfiles); err != nil {
```
commented out tests will need to be undone

mhupman commented on location.go:
```diff
@@ -10,13 +10,14 @@ import (
 	"fmt"
 )
 
+const ENTITYTYPE_LOCATION EntityType = "LOCATION"
```
would suggest leaving this file alone and starting fresh with `entity_location.go` - location service needs to continue working as is

mhupman commented on location_diff.go:
```diff
@@ -28,7 +28,8 @@ type Comparable interface {
 //   // delta -> nil
 func (y Location) Diff(b *Location) (d *Location, diff bool) {
-	d = &Location{Id: String(y.GetId())}
+	d = &Location{EntityMeta: &EntityMeta{
```
similar to above, would suggest leaving this alone and starting with an different differ for Entity usage.  i can't imagine these can share the same code

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+	PageToken string        `json:"pageToken"`
+}
+
+func (e *EntityService) RegisterDefaultEntities() {
```
who calls this?  should probably make it automatic in the client bootstrapping, if possible.  maybe part of the Config() as a toggle?

mhupman replied:
perhaps allowing it to be controlled isn't worth it

cdworak replied:
Config is a little barebones, I'll make it automatic in the bootstrapping and we can always add it to the config later if we need to control it

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+}
+
+func (e *EntityService) Get(id string) (Entity, *Response, error) {
+	var v interface{}
```
could make it a little easier on yourself and make this `map[string]interface{}`

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+		return nil, r, err
+	}
+
+	entity, err := e.toEntityType(v)
```
error handling?

cdworak replied:
added

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+		return nil, fmt.Errorf("Unable to find entityType attribute in %v", metaByKey)
+	}
+
+	entityObj, err := e.LookupEntity(EntityType(entityType.(string)))
```
we should think about what we want to happen if a type is NOT registered.  for example, what if the caller knew they didn't know the types ahead of time and wanted to just inspect them as JSON/map[string]interface{}?  we might want to let that through and not error out in some cases?

mhupman replied:
on the other hand, i can also see it useful to know if you weren't able to unwrap the type, but maybe we find that our during type assertion?

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+	// Convert into struct of Entity Type
+	entityJSON, err := json.Marshal(entityValsByKey)
+	if err != nil {
+		return nil, fmt.Errorf("Error marshaling entity to JSON: %s", err)
```
don't include "error" in error message or capitalize first word

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+
+	err = json.Unmarshal(entityJSON, &entityObj)
+	if err != nil {
+		return nil, fmt.Errorf("Error unmarshaling entity JSON: %s", err)
```
same

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+	return buf.Bytes(), nil
+}
+
+func (e *EntityService) toEntityTypes(entityInterfaces []interface{}) ([]Entity, error) {
```
this is throughout the PR, but the param names are stuttering with their type... prefer shortened values or descriptions of the content without the type:

toEntityTypes(entities []interface{})

cdworak replied:
yeah....i had a hard time because both the word type and interface you can't use in go, so i prefixed them a lot. I'll try to clean it up a bit

mhupman commented on entity_service.go:
```diff
@@ -0,0 +1,253 @@
+	return entities, nil
+}
+
+func (e *EntityService) toEntityType(entityInterface interface{}) (Entity, error) {
```
similarly, `toEntityType(entity interface{})` or `toEntityType(e interface{})`

mhupman reviewed (commented):
nice start!  i agree the Register/Unwrap might need some tweaking, but only usage will tell for sure

mhupman commented on customfield_service.go:
```diff
@@ -531,7 +531,6 @@ func ParseCustomFields(cfraw map[string]interface{}, cfs []*CustomField) (map[st
 }
 
-// TODO: Can we do this anymore?
 func validateCustomFieldsKeys(cfs map[string]interface{}) error {
```
maybe we could rename this to validateLocationCustomFields or something to indicate it is legacy

mhupman commented on registry.go:
```diff
@@ -0,0 +1,30 @@
+
+func (r Registry) Register(key string, val interface{}) {
+	//From: https://github.com/reggo/reggo/blob/master/common/common.go#L169
+	isPtr := reflect.ValueOf(val).Kind() == reflect.Ptr
```
var (..) for all these

mhupman commented on cftasset_service.go:
```diff
@@ -0,0 +1,148 @@
+package yext
```
why is CFT Asset a different service?  can this be part of a unified AssetService?

mhupman commented on cftasset.go:
```diff
@@ -0,0 +1,152 @@
+package yext
```
similarly, isn't this just asset.go?

mhupman commented on asset.go:
```diff
@@ -1,20 +1,11 @@
 package yext
```
High level feedback discussed with @upalsaha:

* Maintain AssetService but rename it to LegacyAssetService
* Name the new service (with registry, entity and CFT support) AssetService
* Add a switch in NewClient() that decides what service to create based on the Version date - leave the other service `nil`

cc @cdworak

mhupman commented on cftasset.go:
```diff
@@ -0,0 +1,152 @@
+	ASSETTYPE_COMPLEXVIDEO AssetType = "complexVideo"
+)
+
+type CFTAsset struct {
```
per above, this would just be Asset - the old thing would become LegacyAsset

mhupman commented on cftasset.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (a *AssetUsage) Equal(b Comparable) bool {
+    defer func() {
```
why is this necessary?

mhupman commented on cftasset.go:
```diff
@@ -0,0 +1,152 @@
+        }
+    }()
+
+    bAssetUsage := *b.(*AssetUsage)
```
is this dereference safe?

mhupman commented on cftasset.go:
```diff
@@ -0,0 +1,152 @@
+type AssetUsageList []AssetUsage
+
+func (a *AssetUsageList) Equal(b Comparable) bool {
+    defer func() {
```
same

mhupman commented on cftasset_diff.go:
```diff
@@ -0,0 +1,21 @@
+	}
+
+	if a.GetAssetType() != b.GetAssetType() {
+		return nil, true
```
nil?

mhupman commented on cftasset_diff.go:
```diff
@@ -0,0 +1,21 @@
+	if a == nil && b != nil {
+		return b, true
+	} else if b == nil && a != nil {
+		return a, true
```
if b is nil there are no proposed changes, right?  wouldn't that be `nil, false`

mhupman commented on cftasset_service.go:
```diff
@@ -0,0 +1,148 @@
+	a.RegisterAssetValue(ASSETTYPE_COMPLEXVIDEO, &ComplexVideoValue{})
+}
+
+func (a *CFTAssetService) RegisterAssetValue(t AssetType, assetValue interface{}) {
```
just curious - any reason we have to wrap these functions vs exposing the Registry directly?

mhupman commented on cftasset_service.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func (a *CFTAssetService) toAssetWithValue(asset *CFTAsset) error {
+	if asset.Type == ASSETTYPE_TEXT {
```
i think i asked this before, but why not just register ASSETTYPE_TEXT so you don't have to special case this?

mhupman commented on customfield.go:
```diff
@@ -231,20 +233,34 @@ func (v *VideoGallery) CustomFieldTag() string {
-	HolidayHours   []HolidayHours `json:"holidayHours,omitempty"`
+// HoursCustom is the Hours custom field format used by locations API
+// Entities API uses the Hours struct in location_entities.go (profile and custom hours are defined the same way for entities)
+type HoursCustom struct {
```
if we follow the other pattern, maybe LegacyHours?

mhupman commented on entity.go:
```diff
@@ -0,0 +1,68 @@
+	b.nilIsEmpty = val
+}
+
+type RawEntity map[string]interface{}
```
another option is to do:

struct RawEntity {
  Meta Meta
  fields: map[string]interface
}

and implement json.Unmarshaler on it.  then you wouldn't need to parse meta by hand

mhupman replied:
yeah i like this

mhupman commented on location_diff.go:
```diff
@@ -148,7 +150,7 @@ func isZeroValue(v reflect.Value, interpretNilAsZeroValue bool) bool {
 	case reflect.Struct:
 		for i, n := 0, v.NumField(); i < n; i++ {
-			if !isZeroValue(v.Field(i), interpretNilAsZeroValue) {
+			if !isZeroValue(v.Field(i), true) {
```
is this a change we want in location_diff

mhupman replied:
(vs entity_diff)

mhupman commented on location_entity.go:
```diff
@@ -0,0 +1,670 @@
+
+func (y LocationEntity) SetLabelIds(v []string) {
+	l := UnorderedStrings(v)
+	y.SetLabelIdsWithUnorderedStrings(l)
```
can you inline the UnorderedStrings wrapper?

cdworak commented on location_entity.go:
```diff
@@ -509,20 +506,20 @@ func (y LocationEntity) GetCategoryIds() (v []string) {
-	if y.LabelIds != nil {
-		v = *y.LabelIds
+func (y LocationEntity) GetLabels() (v UnorderedStrings) {
+	if y.BaseEntity.Meta.Labels != nil {
```
we should probably also move these functions under BaseEntity

xzou committed:
Add LocationType variables

cdworak committed:
**Date to **string

byronharvey24 committed:
added cf endpoint location entitytype

cdworak committed:
hours: Strip leading 0's

byronharvey24 committed:
add external first party reviews create endpoint

byronharvey24 committed:
Merge branch 'createReviewsEndpoint' into entities

alexisgrow committed:
Add lastLoginDate to Users struct PC-43487 (#84)

byronharvey24 committed:
update reviews service to handle response

byronharvey24 committed:
Merge branch 'reviewsUpdates' into entities

xzou committed:
Add GetCountryCode() to LocationEntity

cdworak committed:
customfield_service helpers, add restuarant entity

xzou committed:
Update variables for healthcare professional and healthcare facility

bhainesva committed:
Return Hours reference from SetClosedAllWeek (#89)

* Add NewHoursClosedAllWeek static method

cdworak committed:
add healthcare_professional entity

cdworak committed:
unorderedstrings: if both nil, not diff

byronharvey24 committed:
Create review diff, update pointers, and add getter to entities

pr fixes

cdworak committed:
entity_diff: prevent diff from returning too early

cdworak committed:
Add LocationType Event

xzou committed:
Add methods to DayHours; null check *Hours in GetDayHours

bhainesva committed:
Update searchID param to searchIDs (#96)

alexisgrow committed:
CustomFieldService: stop adding duplicates for multioption select ids

cdworak committed:
modify healthcare professional entity

xzou committed:
Update hours helper struct serialize to return NullRegularHours()

cdworak committed:
export helper functions (#103)

alexisgrow committed:
add entity ACL function to users_service (#100)

bhainesva committed:
Replace LanguageProfile with Entity (#104)

byronharvey24 committed:
Add delete operation to entity service (#105)

* Add delete operation to entity service

* pr fix

* update request to use caps

* rename function

cdworak committed:
language profile upsert: add id param

bhainesva committed:
add support for LanguageProfile:ListAll (#107)

* add support for LanguageProfile:ListAll

Commit:
yext-go changes for MBE ETL Implementation (#108)

Commit:
Update generic diff function name (#109)

xzou committed:
Add methods to get the names of set multi-option and single-option values

cdworak committed:
Set field on generic entity (#111)

cdworak committed:
entity_unmarshal: handle null day hours (and other null values in objects)

xzou committed:
entity_unmarshal: Check if type is struct before calling NumField()

xzou committed:
entity_service: Add filters functionality

cdworak committed:
entity: update json tag of specialties to specialities to match Yext API bug (#115)

tbirdsall committed:
Lat long (#116)

* location: added yextDisplayLat and yextDisplayLng

* location: update spacing

pshah28 committed:
Degrees Fix (#117)

cdworak committed:
customfield_service: handle null option value (#118)

xzou committed:
Change type of logos from **Image to **Photo

bhainesva committed:
Add support for fields param (#119)

cdworak committed:
entity_diff: handle double pointer to nil (aka null)

creotutar committed:
Add Get NPI for Healthcare Professional Entity

alexisgrow committed:
unordered strings equal: Update null checking for interface (#122)

cdworak committed:
Hours: allow for Closed or closed (#124)

pshah28 committed:
Add geomodifier to location entity (#127)

cdworak committed:
SetLabelsWithUnorderedStrings: handle nil J=PC-43989

cdworak committed:
Register healthcare professional entity and add Get functions

cdworak committed:
get and set custom field values by field name

cdworak committed:
updating gender to **string

byronharvey24 committed:
Comment labels

xzou committed:
Add and register ATM entity

xzou committed:
Remove references to upper-snake-case entity types

Done following Product update that has made entity types
returned from the CF endpoint camel case for consistency
with the entities API

byronharvey24 committed:
Merge branch 'entities' of github.com:yext/yext-go into commentLabels

byronharvey24 committed:
Merge branch 'entities' of github.com:yext/yext-go into commentLabels

pshah28 committed:
add healthcare facility entity (#136)

* add healthcare facility entity

* remove todo

byronharvey24 committed:
Merge branch 'commentLabels' into entities

the-nemz committed:
Added custom field type tests to entity differ (#140)

* Added custom field type tests to entity differ

* Fix bug with null types

the-nemz committed:
Add getters for VideoUrl (#141)

cdworak committed:
Diff() for RawEntity

xzou committed:
Add constant for Yext hours format

xzou committed:
Add and register Hotel entity (#138)

bjulian5 committed:
J=PC-59368 Add image width, height and thumbnails fields for images (#146)

Zavoky committed:
Adding breakfast field values (#149)

pshah28 committed:
add google fields to hotel entity (#145)

* add google fields to hotel entity

* modify v2 loc converter to accept hotel

* fix tabbing

pshah28 committed:
add GetMultiOptionIds for error handling (#151)

Zavoky committed:
Add wifi profile field values (#152)

jesuyedavid committed:
Added Insurances Accepted field to healthcare facility (#153)

pshah28 committed:
return diff on nil fields (#154)

cdworak committed:
reorder attributes of location entity

AidanFitzgerald committed:
Add performers field to event entity (#148)

jesuyedavid committed:
hours: unspecified all week (#157)

Zavoky committed:
Adding more hotel profile fields (#156)

* Adding more hotel profile fields

Zavoky committed:
Changing to single pointer strings (#158)

AidanFitzgerald committed:
added reviewLanguage and transactionId to reviews struct

AidanFitzgerald committed:
added more fields to reviews struct

Zavoky committed:
Fixing bicycle values (#161)

Commit:
Updates alongside account syncer implementation (#162)

cdworak committed:
RawEntityClient (#163)

cdworak committed:
customfield create: preserve id (#164)

Zavoky committed:
Adding linked location field to events (#165)

pshah28 committed:
add GetUnorderedStrings helper (#167)

etseff committed:
Add invitationId to Review struct (#169)

Commit:
Improve error messaging for ETLs J=PC-65697 (#170)

Zavoky committed:
Adding timezone field (#171)

AidanFitzgerald committed:
add ReservationUrl to Healthcare Professional entity (#173)

J=none
TEST=manual

tested in Heartland Dental ETL

Commit:
Create helper function for converting string slice to EntityType slice (#174)

cdworak committed:
raw entity: get and set helpers

cdworak committed:
Remove WheelchairAccessbile (same as MobilityAccessible)

cdworak committed:
hotel: add all inclusive options

Zavoky committed:
Delete entity ID before upsert (#178)

cdworak committed:
language profile service: handle nil meta

jesuyedavid committed:
update yext menu fields (#180)

cdworak committed:
hours: rename to location_hours to signify that is legacy format (#181)

Zavoky committed:
changing calorie type to *int

cases where calorie was intentionally set to 0 were being omitted

jesuyedavid committed:
add job entity (#182)

Zavoky committed:
Adding Facebook cover & profile fields (#183)

Commit:
Add Geomodifier to Restaurant Entity Definition (#184)

AidanFitzgerald committed:
add facebookcta field and constants

AidanFitzgerald committed:
-J=PC-63853 changes to entity_registry and unmarshal needed for hybrid history mode

AidanFitzgerald committed:
fix unnecessary double ifs

AidanFitzgerald committed:
Merge branch 'entities-hhm' into entities

pshah28 committed:
prevent hours parsing panic and return error (#190)

Commit:
analytics_service: Always return the response and error (#191)

TEST=manual

Commit:
Add definition for FAQ entity type

pshah28 committed:
add reopen date to hours (#192)

cdworak committed:
Add format param

Zavoky committed:
adding fields to analytics request (#194)

Zavoky committed:
Analytics fields (#195)

* adding fields to analytics request

* Adding EntityType to analytics filter

AidanFitzgerald committed:
add education type constants for healthcareprofessionals

nsiskind committed:
customfields: adding validation field (#197)

validation field allows for rich text formatting
options to be unmarshalled by api request.
Necessary for enhancing custom field create
functionality in ETLs.

J=PC-78670
TEST=manual

nsiskind committed:
customfield.go : update custom field validation min/max val to float64 (#199)

nirmalpatel94 committed:
feat(healthcare): add telehealth url field to providers and facilities

J=PC-81848

nirmalpatel94 committed:
feat(hotel): add indoor and outdoor pool fields to hotel entities

J=PC-81624

nirmalpatel94 committed:
feat(healthcare): add getter for display url for telehealth url field

nirmalpatel94 committed:
fix(healthcare): fix telehealth type definition

stephh2 committed:
entity location: add 'black owned business' and 'pickup and delivery service' profile field

nirmalpatel94 committed:
feat(jobs): add Location profile field to job struct

TEST=manual

nirmalpatel94 committed:
fix(jobs): fix api name for application url

TEST=manual

wenright committed:
feat(entity): add service area places (#205)

danielwuhoo committed:
feat(hotel): added pools to hotel entities (#206)

J=PC-108774

Zavoky committed:
adding 'accepting new patients' and 'covid 19 info url' (#207)

AidanFitzgerald committed:
feat(entity): adds RemoveEntry for rawEntity (#208)

* feat(entity): adds RemoveEntry for rawEntity

J=PC-112865
TEST=manual

* refactor(entity): removes print statements

J=PC-112865
TEST=manual

christianterence committed:
feat(entity-location): add SourceUrl as an exposed string for struct type Image for yext.Photo

J=PC-121492
TEST=manual

xzou committed:
feat: add drive-through hours profile field

Add drive-through hours to location and ATM entities

J=PC-122472
TEST=manual

nsiskind committed:
entity_location: add cta yext type (#212)

* entity_location: add cta yext type

* *feat(entity_location): add nullableCTA to cta yext type

Commit:
Update analytics review fields to match what was changed in this CR https://gerrit.yext.com/c/alpha/+/125658
TEST=manual

nirmalpatel94 committed:
feat(entity_service): add rich text format field

TEST=manual

Commit:
Remove publisherSuggestionType, searchTerm, searchType, and foursquareCheckinTimeOfDay from AnalyticsFilters
J=PC-126903
TEST=auto

AidanFitzgerald committed:
adds accessHours profile field to locations

AidanFitzgerald committed:
adds new func to determine if raw entity is zero value

AidanFitzgerald committed:
add tests and support for []interface{}

AidanFitzgerald committed:
adds funcs for getting all option names for option select fields

AidanFitzgerald committed:
fixes bug in rawEntityIsZeroValue

nsiskind committed:
adds helpArticle entity type (#220)

* adds helpArticle entity type

* entity_helparticle: use better var names

Co-authored-by: afitzgerald <afitzgerald@yext.com>

nsiskind committed:
location_customfield: add ToDate function for yext.Date (#221)

* location_customfield: add ToDate function for yext.Date

J=PC-127422

* type.go: add toDate function and remove from location customfield

J=PC-127422

nsiskind committed:
helparticle: add landing page URL field (#222)

zxu10 committed:
Add Activity Logs and update some Analytics Report (#223)

* Add Activity Logs Struct and Service

J=PC-134868

* Update some Analytics Data API names

J=PC-135005

AidanFitzgerald committed:
adds DigitalGuestRoomKeys field to hotel entity struct (#224)

zxu10 committed:
add activitylogservice to client.go (#225)

zxu10 committed:
update actordetails to activity log (#226)

JamesJMH committed:
Add google place id (#227)

zxu10 committed:
update activity log and analytics data for att J=PC-134030 TEST=manual (#228)

Commit:
healthcare professional: implements setValidDegrees method

- Checks degrees against hardcoded list of valid degrees supported by our platform
- sets degrees on self, returns parse error if any are invalid

J=PC-113595
TEST=manual

zxu10 committed:
fix activity log email (#230)

JamesJMH committed:
Add googlePlaceId for healthcare professionals (#231)

zxu10 committed:
fix page views api name and add cumulative rating get function (#233)

JamesJMH committed:
add npi for healthcare facilities (#234)

Commit:
feat: adds profile field "Impressum" to location entity struct (#235)

J=PC-130370
TEST=none

zxu10 committed:
add competitor field to analytics data (#236)

aliang97 committed:
feat(entity): add new entity type "Financial Professional"

J=PC-143030
TEST=manual

AidanFitzgerald committed:
adds covid fields to hotel entity type

John-Nguyen committed:
Added delivery hours field to restaurant entity

J=PC-146897 (#240)
TEST=manual

marclanepitt committed:
adds googleMyBusinessLabels to healthcare professional and facility (#239)

* adds googleMyBusinessLabels to healthcare professional and facility

J=PC-146073
TEST=manual

Ran test write api run with steward etl

* use unordered strings

Commit:
feat(location_entity): adds timezone field to entity (#242)

J=PC-27188
TEST=Manual

marclanepitt committed:
add verbose response logging when we encounter a failure in our Do (#243)

request logic
Goal is to help debug unhelpful 'unexpected end of JSON input' logs

J=CR-2054
TEST=auto

nsiskind committed:
entity_location: make GoogleWebsiteOverride a **string (#245)

* entity_location: make GoogleWebsiteOverride a **string

* Make GoogleWebsiteOverride a string across all entity types

nsiskind committed:
Googlewebsite fix (#247)

* entity_location: make GoogleWebsiteOverride a **string

* Make GoogleWebsiteOverride a string across all entity types

* update GoogleWebsiteOverride to be **string
for healthcare entity types

* update GoogleWebsiteOverride to be **string for all entityy types

marclanepitt committed:
add request body to err message for debugging request failures (#246)

J=CR-2054
TEST=auto

marclanepitt committed:
fix, update formatted string to use string (#248)

Zavoky committed:
additions to valid degrees & job field (#244)

marclanepitt committed:
print dump response error (#249)

J=CR-2054
TEST=auto

jpjamipark committed:
fix typo in hiringOrganization json

stephh2 committed:
Add analytics clicks event and update page views event (#252)

stephh2 committed:
Add Entity ID metric to analytics data struct (#253)

Zavoky committed:
client.go error logging (#251)

nsiskind committed:
price: add price object (#255)

marclanepitt committed:
retry network request on connection reset from server (#256)

J=CR-2110
TEST=manual

jpjamipark committed:
feat(reviews): add reviews update endpoint

stephh2 committed:
Add label ids as review request param

danielwuhoo committed:
feat: add service for admin apis (#254)

Zavoky committed:
nil platform bool != zero value (#259)

* nil platform bool != zero value

when dealing with yes/no fields in the Yext platform, there is also an unspecified value (null). This change accounts for cases where we want to set that unspecified value to 'No'

Zavoky committed:
Revert "nil platform bool != zero value (#259)"

This reverts commit f8b7e7c4efe25117b9513ec325b6f4da26fdc149. Reverting until I can properly test the results

danielwuhoo committed:
feat: add services service for add requests (#260)

https://hitchhikers.yext.com/docs/adminapis/addrequests/

jpjamipark committed:
feat: add FAQ fields to locationEntity

Zavoky committed:
add 'strip unsupported formats' (#262)

jpjamipark committed:
add new answers values to analytics structs

cdworak committed:
review-service: add UpdateLabels, externalId (#263)

* review-service:add review update

* review: add externalId field

cdworak committed:
review: removing duplicate external id (#264)

cdworak committed:
review-service: add external id to update object (#265)

cblair28 committed:
add Entity struct to Reviewer struct (#267)

Co-authored-by: sharvey <sharvey@yext.com>

marclanepitt committed:
add pre processing option ot menu list data (#268)

needed for subway etl to cut down on memory usage

J=PC-122295
TEST=manual

sri-jayakumar committed:
custom-field-service: add translations field to options (#269)

marclanepitt committed:
update service existinglocationaddresponse type for agreementid (#271)

mtusman committed:
add rendered option for entity language profiles (#270)

AidanFitzgerald committed:
adds GoogleMyBusinessLabels field to location entities (#272)

TBRing

AidanFitzgerald committed:
adds pickupdelivery field

Zavoky committed:
added employmentType field to job entities

Zavoky committed:
added pickuphours to location entity

jesuyedavid committed:
rnd: add progress bar to show entity loading from API progress (#273)

nirmalpatel94 committed:
fix(entity_service): add nil check

J=CO-2055
TEST=manual

jesuyedavid committed:
entity_service: add hide progress bar option (#275)

hoosRed committed:
Add Geomodifier to Healthcare Professional entity

NathanielGE committed:
making a user's createdDate field accessible (#277)

JohnCho92 committed:
add brunch hours and drive-through hours to restaurant entity (#279)

J=PC-155508

Co-authored-by: JohnCho92 <jcho@yext.com>

Zavoky committed:
added option to consider nil booleans as non-zero values (#281)

when nilIsEmpty is set to true, the differ considers nil booleans as the zero value ("false"). This commit introduces nilBoolIsEmpty, a flag specifically for booleans that can override nilIsEmpty

Zavoky committed:
updated logic for **bools found in pointers

AidanFitzgerald committed:
Entities facebookusername (#282)

* adds facebookusername field for location entities

* fixes formatting??

JohnCho92 committed:
add Slug to Healthcare Professional entity (#283)

Co-authored-by: JohnCho92 <jcho@yext.com>

JohnCho92 committed:
add Slug to Healthcare Facility entity (#284)

Co-authored-by: JohnCho92 <jcho@yext.com>

Commit:
PC-182867: add google entity relationship location entity field (#285)

* PC-182867: add google entity relationship location entity field

* make it double pointer

nsiskind committed:
Update Financial Professional Field types for DisclosureLink AndroidAppURL IOSAppURL LandingPageUrl (#286)

mtusman committed:
add invitationId to ReviewCreateInvitationResponse struct (#287)

cdworak committed:
change isclosed from **bool to bool to treat nil and false the same (#288)

nsiskind committed:
Listings (#289)

* add listings service

* add list all endpoint

* fix listing response obj

AidanFitzgerald committed:
adds entity_product.go for product entity type (#291)

mtusman committed:
modify AnalyticsData struct (#292)

Zavoky committed:
Adding Google Place ID & Facebook Store ID

AidanFitzgerald committed:
adds PrimaryPhoto field to products (#293)

* adds PrimaryPhoto field to products

* adds landingpageurl too

jesuyedavid committed:
add createreviewliveapi to review service (#294)

* add createreviewliveapi to review service

* use double pointer for review entity struct

mtusman committed:
add support for listing impressions metric (#295)

matthewleifer committed:
add rankTrackingCompetitors to location entity

AidanFitzgerald committed:
add timezone field for restaurant entities (#297)

kyled526 committed:
feat(custom): amend pr comments:wq

to entity healthcare professional/healthcare facility for ETL use

J=PC-213266
TEST=manual

JohnCho92 committed:
add facebook location descriptor and username fields for restaurant entities (#299)

Co-authored-by: JohnCho92 <jcho@yext.com>

nirmalpatel94 committed:
feat: add custom unmarshalling for newer V param's custom fields

- newer API versions return custom fields differently. This change
  will give some backwards compatibility

J=PC-219314
TEST=manual

mtusman committed:
migrate custom field endpoint to optionally use the configuration end… (#301)

* migrate custom field endpoint to optionally use the configuration endpoint

J=CR-3126
TEST=auto

* refactor: respond to dev comments

J=CR-3126

kunal-p committed:
feat: add config option to use EU host domain for yextapis.com (#302)

junpark1213 committed:
fix: update AnalyticsFilters FolderId name and type

J=PC-220346

junpark1213 committed:
fix: update click metric to new click count metric

stephh2 committed:
feat(hotel & restaurant): add additional fields to base struct (#304)

J=none
TEST=manual

Co-authored-by: sseverance <sseverance@yext.com>

Commit:
add product lists to financial professional entities (#306)

J=PC-234251
TEST=auto

junpark1213 committed:
fix: remove deprecated metrics

J=PC-220346

dgurland committed:
fix(restaurant): add DineInHours to base struct

J=PC-237814
TEST=manual

nirmalpatel94 committed:
feat(reviews): add new review api fields

- https://yexttest.atlassian.net/browse/PB-21550

J=PC-233325
TEST=manual

Jeffrey-Rhoads17 committed:
feat(custom): add nonbinary gender to healthcare professional entity (#310)

J=PC-240725
TEST=manual

2018lwang committed:
feat(custom): change supported format struct for richtextv2 field (#311)

J=PC-240487
TEST=manual

Co-authored-by: Lilian Wang <lwang@yext.com>

jknutsonn committed:
feat: add notifyUser field to user

J=PC-244952

dkang3021 committed:
feat(entity_service): add convertRichTextToHTML param (#313)

J=PC-244921
TEST=manual

peteros086 committed:
feat(hotel+restaurant): add addl fields (#314)

J=none|TEST=manual

Jeffrey-Rhoads17 committed:
fix(custom): allow for multiple data types of IDs from Review Update endpoint to account for newer A (#315)

J=PC-254056
TEST=manual

Jeffrey-Rhoads17 committed:
Revert "feat(hotel+restaurant): add addl fields (#314)" (#316)

This reverts commit 919a4bf47eb2f0f04a06d5ebc649cb093ae72da4.

Jeffrey-Rhoads17 committed:
feat(hotel+restaurant): add addl fields (#317)

J=none|TEST=manual

Co-authored-by: Peter Oliveira Soens <poliveirasoens@yext.com>

peteros086 committed:
Revert "feat(hotel+restaurant): add addl fields (#317)" (#318)

This reverts commit d453c8db678179a900ddff08cab9c5ea3b4764c3.

matthewleifer committed:
feat(assets): add folderids field

J=none

Commit:
feat(entities): add timestamp field to entity-meta

J=PC-258895

Commit:
feat(entities): add getter function for timestamp field

J=PC-258895

Commit:
feat(entities): add GetTimestampAsTime function

J=PC-258895

Commit:
Revert "feat(entities): add GetTimestampAsTime function"

This reverts commit d840ea08cbd6537bebaa1cf8037be29bfc44a29c.

nirmalpatel94 committed:
chore: add instructions for making changes and deploying`

stephh2 committed:
feat(license): add license service (#323)

J=none

kaneleeyext committed:
feat(license): fix passing opts to tokenListHelper (#324)

J=none

tritpham committed:
feat(healthcare-professional): Add YextDisplayCoordinate property and Get functions (#325)

nirmalpatel94 committed:
Update README.md

fix command

stephh2 committed:
license-service: list license assignments in single object (#326)

Co-authored-by: Stephanie Severance <sseverance@yext.com>

ajgist committed:
entity: add CreatedTimestamp field to entityMeta struct (#327)

Jared-Hood committed:
client: add optional Update-Operation-Id header to API requests (#328)

Jeffrey-Rhoads17 committed:
fix(custom): add option for list CF types and proper fallback for CFTs to set the type correctly (#329)

J=PC-270239
TEST=manual

bungaepyo committed:
fix(analytics): update ResponseRate type from int to float

PC-279430

dgurland committed:
analytics service: remove deprecated fields
based on https://gerrit.yext.com/c/alpha/+/342625
J=PC-255270
TEST=manual
